PolySTAR-Taproot-project/src/communication/cv_serial.cpp

```diff
@@ -0,0 +1,125 @@
+#include "tap/architecture/clock.hpp"
+#include "tap/architecture/endianness_wrappers.hpp"
+#include "tap/communication/serial/uart.hpp"
+#include "tap/drivers.hpp"
+#include "tap/errors/create_errors.hpp"
+
+#include "cv_serial.hpp"
+
+/**
+ * Macro that wraps uart read for ease of readability in code.
+ * @param[in] data Byte array where the data should be read into.
+ * @param[in] length The number of bytes to read.
+ * @return The number of bytes read into data.
+ */
+#define READ(data, length) drivers->uart.read(this->port, data, length)
+
+namespace src::communication::cv
+{
+CVSerial::CVSerial(Drivers *drivers, Uart::UartPort port)
+    : port(port),
+      cvSerialRxState(SERIAL_HEADER_SEARCH),
+      newMessage(),
+      mostRecentMessage(),
+      frameCurrReadByte(0),
+      drivers(drivers)
+{
+}
+
+void CVSerial::initialize()
+{
+    switch (this->port)
+    {
+        case Uart::UartPort::Uart1:
+            drivers->uart.init<Uart::UartPort::Uart1, 230400>();
+            break;
+        case Uart::UartPort::Uart2:
+            drivers->uart.init<Uart::UartPort::Uart2, 230400>();
+            break;
+        case Uart::UartPort::Uart3:
+            drivers->uart.init<Uart::UartPort::Uart3, 230400>();
+            break;
+        case Uart::UartPort::Uart6:
+            drivers->uart.init<Uart::UartPort::Uart6, 230400>();
+            break;
+        case Uart::UartPort::Uart7:
+            drivers->uart.init<Uart::UartPort::Uart7, 230400>();
+            break;
+        case Uart::UartPort::Uart8:
+            drivers->uart.init<Uart::UartPort::Uart8, 230400>();
+            break;
+        default:
+            break;
+    }
+}

```

Comment:
i'm not sure about this but if the switch is just to avoid 2 initializing specific cases, maybe one if would do the job since the initialization is always basically the same and just uses the port ?

Reply:
I'm guessing this is because of the fact the initialization is template-dependant, so you have to explicitly call the functions themselves rather than as an argument. In that case there isn't anything else to do than this switch statement. Why are you excluding 4 & 5 though @FreshOlives  ?

Reply:
I need to check to be completely sure, but I'm fairly confident that's because UART 4 and 5, while they technically exist on the STM chip, are not accessible on the Board A